### PR TITLE
Add pub cache, artifacts, pkgs to Cirrus cache

### PR DESCRIPTION
**Pull request by DanTup**
This adds:

- pub's cache to the Cirrus cache. It caches based on the checksums in the pubspec (although pub has things in versioned folders, this'll prevent the cache just getting bigger over time).
- bin/cache/pkg. I think this only includes `sky_engine`. It caches based on the engine checksum.
- bin/cache/artifacts. This includes `engine`, `gradle_wrapper`, `material_fonts`. Also caches based on the engine checksum.

### CODE REVIEW

**Comment by DanTup** [comment]
Note: Some of this duplication will go away with https://github.com/flutter/flutter/pull/20077 (I'll rebase and resolve conflicts once one of them lands).

File: .cirrus.yml
Comment (gspencergoog) [inline .cirrus.yml]: Use `$CIRRUS_WORKING_DIR` instead of `.`, just to be more explicit about where it's executing.

Also, this would be more efficient as a pipe to xargs:
```bash
find "$CIRRUS_WORKING_DIR" -name pubspec.yaml | xargs grep -H 'PUBSPEC CHECKSUM'
```
-exec invokes a new grep for each item found, but xargs will invoke grep on as many files as it can at once, which is more efficient.  In fact, the following is even more efficient:
```bash
grep -r --include=pubspec.yaml 'PUBSPEC CHECKSUM' "$CIRRUS_WORKING_DIR"
```
Since it only invokes one executable once.

File: .cirrus.yml
Comment (gspencergoog) [inline .cirrus.yml]: (same comment as above)

File: .cirrus.yml
Comment (gspencergoog) [inline .cirrus.yml]: (same as above)

File: .cirrus.yml
Comment (DanTup) [reply .cirrus.yml]: I read that the `+` on the end of find would pass all filenames once, but I'm not familiar with these so I don't know if it's exactly the same.

I'll change to what you suggest last though - I did look for how to do it all with grep and somehow failed to find that!

(I'll wait until I've landed the other change before fixing this, so it's in less places, and it's going to conflict anyway).

**Review by gspencergoog — APPROVED** [review]

**Comment by gspencergoog** [comment]
(modulo the grep changes)
32384589-a60f0e74-c078-11e7-9bc1-e5b5287aea9d

**Comment by DanTup** [comment]
I've rebased to remove some of the dupes (the grep command is still there twice, linux and mac for now) and made the requested changes.

**Comment by DanTup** [comment]
I'm slightly suspicious about this now... It looks to have sped the build scripts up, but uploading the cache seems slow. For example:

https://cirrus-ci.com/task/6426244023320576

upload_artifacts_cache took a minute and a half, which offsets the benefit. If this is normal, this caching might not be as helpful as hoped. I'll ping Cirrus to see if this is normal before I consider landing this.

**Comment by DanTup** [comment]
Ok, looks like these caches aren't per-OS, so our builds are overwriting each others caches (the artifacts folder is different on each platform). I'll add the OS into the fingerprint and see how that affects things.

**Comment by DanTup** [comment]
@gspencergoog I've added the OS to the fingerprint because the different OSes were overwriting each others artifact cache on every build. Please check the new changesets.

The Linux and Windows builds that are running now should show cache hits and then *not* upload the caches at the end (macOS will upload, because it didn't get a previous completed build with the OS in the fingerprint). If that's the case, this should be good to land, and hopefully will be faster than master.

I'm collecting some notes on things we may be able to speed up in https://github.com/flutter/flutter/issues/20036.

**Comment by gspencergoog** [comment]
Looks like this worked!  Since you're out for the next two weeks, I'll land this on green.

**Comment by DanTup** [comment]
Thanks - some of the improvements were less than I hoped, but I think we can keep chipping away. If we can get the setup stages really fast, we could shard the tests more and bring down the time from opening a PR(/landing a change) to seeing the results.

**Comment by gspencergoog** [comment]
Yeah, my Docker image change should make a big difference on Linux, at least.  Conceivably, we can also do that for Windows, but it'll have to be someone who has a Windows box that sets that up...

**Pull request merged**
